File: internal/interfaces/services/network/dhcp.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package networkServiceInterfaces

import networkModels "github.com/alchemillahq/sylve/internal/db/models/network"

type ModifyDHCPConfigRequest struct {
	StandardSwitches []uint   `json:"standardSwitches"`
	ManualSwitches   []uint   `json:"manualSwitches"`
	DNSServers       []string `json:"dnsServers"`
	Domain           string   `json:"domain"`
	ExpandHosts      *bool    `json:"expandHosts"`
}

type CreateDHCPRangeRequest struct {
	Type           string `json:"type" binding:"required,oneof=ipv4 ipv6"`
	StartIP        string `json:"startIp"`
	EndIP          string `json:"endIp"`
	StandardSwitch *uint  `json:"standardSwitch"`
	ManualSwitch   *uint  `json:"manualSwitch"`
	Expiry         *uint  `json:"expiry"`
	RAOnly         *bool  `json:"raOnly"`
	SLAAC          *bool  `json:"slaac"`
}

type ModifyDHCPRangeRequest struct {
	ID             uint   `json:"id"`
	Type           string `json:"type" binding:"required,oneof=ipv4 ipv6"`
	StartIP        string `json:"startIp"`
	EndIP          string `json:"endIp"`
	StandardSwitch *uint  `json:"standardSwitch"`
	ManualSwitch   *uint  `json:"manualSwitch"`
	Expiry         *uint  `json:"expiry"`
	RAOnly         *bool  `json:"raOnly"`
	SLAAC          *bool  `json:"slaac"`
}

type FileLeases struct {
	Expiry   uint64 `json:"expiry"`
	MAC      string `json:"mac"`
	IAID     string `json:"iaid"`
	IP       string `json:"ip"`
	Hostname string `json:"hostname"`
	ClientID string `json:"clientId"`
	DUID     string `json:"duid"`
}

type Leases struct {
	File []FileLeases                    `json:"file"`
	DB   []networkModels.DHCPStaticLease `json:"db"`
}

type CreateStaticMapRequest struct {
	Hostname     string `json:"hostname"`
	Comments     string `json:"comments"`
	IPObjectID   *uint  `json:"ipId"`
	MACObjectID  *uint  `json:"macId"`
	DUIDObjectID *uint  `json:"duidId"`
	DHCPRangeID  uint   `json:"dhcpRangeId" binding:"required"`
}

type ModifyStaticMapRequest struct {
	ID           uint   `json:"id" binding:"required"`
	Hostname     string `json:"hostname"`
	IPObjectID   *uint  `json:"ipId"`
	MACObjectID  *uint  `json:"macId"`
	DUIDObjectID *uint  `json:"duidId"`
	DHCPRangeID  uint   `json:"dhcpRangeId" binding:"required"`
	Comments     string `json:"comments"`
}

File: web/src/lib/api/network/dhcp.ts
import { APIResponseSchema, type APIResponse } from '$lib/types/common';
import {
	DHCPConfigSchema,
	DHCPRangeSchema,
	LeasesSchema,
	type DHCPConfig,
	type DHCPRange,
	type Leases
} from '$lib/types/network/dhcp';
import { apiRequest } from '$lib/utils/http';
import { z } from 'zod/v4';

export async function getDHCPConfig(): Promise<DHCPConfig> {
	return await apiRequest('/network/dhcp/config', DHCPConfigSchema, 'GET');
}

export async function updateDHCPConfig(
	standardSwitches: number[],
	manualSwitches: number[],
	dnsServers: string[],
	domain: string,
	expandHosts?: boolean
): Promise<APIResponse> {
	const body: {
		standardSwitches: number[];
		manualSwitches: number[];
		dnsServers: string[];
		domain: string;
		expandHosts?: boolean;
	} = {
		standardSwitches,
		manualSwitches,
		dnsServers,
		domain,
		expandHosts
	};

	return await apiRequest('/network/dhcp/config', APIResponseSchema, 'PUT', body);
}

export async function getDHCPRanges(): Promise<DHCPRange[]> {
	return await apiRequest('/network/dhcp/range', z.array(DHCPRangeSchema), 'GET');
}

export async function createDHCPRange(
	ipType: 'ipv4' | 'ipv6',
	startIp: string,
	endIp: string,
	expiry: number,
	raOnly: boolean,
	slaac: boolean,
	standardSwitchId?: number,
	manualSwitchId?: number
): Promise<APIResponse> {
	return await apiRequest('/network/dhcp/range', APIResponseSchema, 'POST', {
		type: ipType,
		startIp,
		endIp,
		expiry,
		standardSwitch: standardSwitchId,
		manualSwitch: manualSwitchId,
		raOnly,
		slaac
	});
}

export async function updateDHCPRange(
	type: 'ipv4' | 'ipv6',
	id: number,
	startIp: string,
	endIp: string,
	expiry: number,
	raOnly: boolean,
	slaac: boolean,
	standardSwitchId?: number,
	manualSwitchId?: number
): Promise<APIResponse> {
	return await apiRequest(`/network/dhcp/range/${id}`, APIResponseSchema, 'PUT', {
		type,
		id,
		startIp,
		endIp,
		expiry,
		raOnly,
		slaac,
		standardSwitch: standardSwitchId,
		manualSwitch: manualSwitchId
	});
}

export async function deleteDHCPRange(id: number): Promise<APIResponse> {
	return await apiRequest(`/network/dhcp/range/${id}`, APIResponseSchema, 'DELETE');
}

export async function getLeases(): Promise<Leases> {
	return await apiRequest('/network/dhcp/lease', LeasesSchema, 'GET');
}

export async function createDHCPLease(
	hostname: string,
	comments: string,
	ipObjectId: number | null,
	macObjectId: number | null,
	duidObjectId: number | null,
	dhcpRangeId: number
): Promise<APIResponse> {
	return await apiRequest('/network/dhcp/lease', APIResponseSchema, 'POST', {
		hostname,
		comments,
		ipId: ipObjectId,
		macId: macObjectId,
		duidId: duidObjectId,
		dhcpRangeId
	});
}

export async function modifyDHCPLease(
	id: number,
	hostname: string,
	comments: string,
	ipObjectId: number | null,
	macObjectId: number | null,
	duidObjectId: number | null,
	dhcpRangeId: number
): Promise<APIResponse> {
	return await apiRequest(`/network/dhcp/lease`, APIResponseSchema, 'PUT', {
		id: Number(id),
		hostname,
		comments,
		ipId: ipObjectId,
		macId: macObjectId,
		duidId: duidObjectId,
		dhcpRangeId
	});
}

export async function deleteDHCPLease(id: number): Promise<APIResponse> {
	return await apiRequest(`/network/dhcp/lease/${id}`, APIResponseSchema, 'DELETE');
}
